### PR TITLE
(enhancement)crds: add openapi specification based schema validation for chaos CRDs

### PR DESCRIPTION
**Pull request by ksatchit**
Signed-off-by: ksatchit <ksatchit@mayadata.io>

**What this PR does / why we need it**:

- The chaosExperiment & chaosEngine CRDs are dev/user facing custom resources in LitmusChaos. These determine the execution and outcome of the chaos experiment & thus need validation at a pre-deployment level to prevent unnecessary runtime failures

- The chaos-executor (esp. the ansible executor) today has complex validation logic that is not needed provided the CRD schema is strictly enforced. One way to achieving this is with the OpenAPI specification based schema validation.  

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #https://github.com/litmuschaos/litmus/issues/1002

**Special notes for your reviewer**:

- ~~This is a WIP PR with validation adding for chaosExperiment CRDs without specifying the "pruning" functionality of the OpenAPI to drop unspecified fields.~~ 

- ~~Support for configmap/secret validation and validation spec for chaosEngine schema is being added in subsequent commits.~~

- The support for ChaosEngine CRD validation with pattern matching for spec components as well as configmap/secret volume support for chaosexpriments (with support to detect malformed spec, i.e., empty params) has been added. 

**Checklist:**
- [ ] Fixes #<issue number>
- [ ] Labelled this PR & related issue with `documentation` tag
- [ ] PR messages has document related information
- [ ] Labelled this PR & related issue with `breaking-changes` tag
- [ ] PR messages has breaking changes related information
- [ ] Labelled this PR & related issue with `requires-upgrade` tag
- [ ] PR messages has upgrade related information
- [ ] Commit has unit tests
- [ ] Commit has integration tests

### CODE REVIEW

**Comment by ksatchit** [comment]
Example validation errors upon `kubectl apply` with incorrect schema (chaosExperiment): 

```
The ChaosExperiment "kafka-broker-network-latency" is invalid: []: Invalid value: map[string]interface {}{"apiVersion":"litmuschaos.io/v1alpha1", "description":map[string]interface {}{"message":"Deleting a kafka broker pod\n"}, "kind":"ChaosExperiment", "metadata":map[string]interface {}{"annotations":map[string]interface {}{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"litmuschaos.io/v1alpha1\",\"description\":{\"message\":\"Deleting a kafka broker pod\\n\"},\"kind\":\"ChaosExperiment\",\"metadata\":{\"annotations\":{},\"generation\":1,\"name\":\"kafka-broker-network-latency\",\"namespace\":\"default\"},\"spec\":{\"definition\":{\"args\":[\"-c\",\"ansible-playbook ./experiments/kafka/kafka-broker-network-latency/kafka-broker-network-latency-ansible-logic.yml -i /etc/ansible/hosts -vv; exit 0\"],\"command\":\"/bin/bash\",\"env\":[{\"name\":\"ANSIBLE_STDOUT_CALLBACK\",\"value\":\"default\"},{\"name\":\"KAFKA_KIND\",\"value\":\"statefulset\"},{\"name\":\"KAFKA_LIVENESS_STREAM\",\"value\":\"enabled\"},{\"name\":\"KAFKA_LIVENESS_IMAGE\",\"value\":\"litmuschaos/kafka-client:ci\"},{\"name\":\"KAFKA_CONSUMER_TIMEOUT\",\"value\":\"30000\"},{\"name\":\"TOTAL_CHAOS_DURATION\",\"value\":\"60000\"},{\"name\":\"TARGET_CONTAINER\",\"value\":\"k8skafka\"},{\"name\":\"NETWORK_LATENCY\",\"value\":\"60000\"},{\"name\":\"NETWORK_INTERFACE\",\"value\":\"eth0\"},{\"name\":\"KAFKA_INSTANCE_NAME\",\"value\":\"\"},{\"name\":\"KAFKA_NAMESPACE\",\"value\":\"\"},{\"name\":\"KAFKA_LABEL\",\"value\":\"\"},{\"name\":\"KAFKA_BROKER\",\"value\":\"\"},{\"name\":\"KAFKA_REPLICATION_FACTOR\",\"value\":\"\"},{\"name\":\"KAFKA_SERVICE\",\"value\":\"\"},{\"name\":\"KAFKA_PORT\",\"value\":\"\"},{\"name\":\"ZOOKEEPER_NAMESPACE\",\"value\":\"\"},{\"name\":\"ZOOKEEPER_LABEL\",\"value\":\"\"},{\"name\":\"ZOOKEEPER_SERVICE\",\"value\":\"\"},{\"name\":\"ZOOKEEPER_PORT\",\"value\":\"\"},{\"name\":\"LIB\",\"value\":\"pumba\"},{\"name\":\"LIB_IMAGE\",\"value\":\"gaiaadm/pumba:0.6.5\"}],\"image\":\"ksatchit/ansible-runner:kafka\",\"labels\":{\"name\":\"kafka-broker-network-latency\"},\"permissions\":{\"apiGroups\":[\"\",\"extensions\",\"apps\",\"batch\",\"litmuschaos.io\"],\"resources\":[\"daemonsets\",\"statefulsets\",\"deployments\",\"replicasets\",\"jobs\",\"pods\",\"pods/exec\",\"chaosengines\",\"chaosexperiments\",\"chaosresults\"],\"verbs\":[\"*\"]}}}}\n"}, "creationTimestamp":"2019-12-19T04:58:34Z", "generation":2, "name":"kafka-broker-network-latency", "namespace":"default", "resourceVersion":"15900770", "uid":"db81b603-8c70-42fa-9cfa-5e1656d57302"}, "spec":map[string]interface {}{"definition":map[string]interface {}{"args":[]interface {}{"-c", "ansible-playbook ./experiments/kafka/kafka-broker-network-latency/kafka-broker-network-latency-ansible-logic.yml -i /etc/ansible/hosts -vv; exit 0"}, "command":"/bin/bash", "env":[]interface {}{map[string]interface {}{"name":"ANSIBLE_STDOUT_CALLBACK", "value":"default"}, map[string]interface {}{"name":"KAFKA_KIND", "value":"statefulset"}, map[string]interface {}{"name":"KAFKA_LIVENESS_STREAM", "value":"enabled"}, map[string]interface {}{"name":"KAFKA_LIVENESS_IMAGE", "value":"litmuschaos/kafka-client:ci"}, map[string]interface {}{"name":"KAFKA_CONSUMER_TIMEOUT", "value":"30000"}, map[string]interface {}{"name":"TOTAL_CHAOS_DURATION", "value":"60000"}, map[string]interface {}{"name":"TARGET_CONTAINER", "value":"k8skafka"}, map[string]interface {}{"name":"NETWORK_LATENCY", "value":"60000"}, map[string]interface {}{"name":"NETWORK_INTERFACE", "value":"eth0"}, map[string]interface {}{"name":"KAFKA_INSTANCE_NAME", "value":""}, map[string]interface {}{"name":"KAFKA_NAMESPACE", "value":""}, map[string]interface {}{"name":"KAFKA_LABEL", "value":""}, map[string]interface {}{"name":"KAFKA_BROKER", "value":""}, map[string]interface {}{"name":"KAFKA_REPLICATION_FACTOR", "value":""}, map[string]interface {}{"name":"KAFKA_SERVICE", "value":""}, map[string]interface {}{"name":"KAFKA_PORT", "value":""}, map[string]interface {}{"name":"ZOOKEEPER_NAMESPACE", "value":""}, map[string]interface {}{"name":"ZOOKEEPER_LABEL", "value":""}, map[string]interface {}{"name":"ZOOKEEPER_SERVICE", "value":""}, map[string]interface {}{"name":"ZOOKEEPER_PORT", "value":""}, map[string]interface {}{"name":"LIB", "value":"pumba"}, map[string]interface {}{"name":"LIB_IMAGE", "value":"gaiaadm/pumba:0.6.5"}}, "image":"ksatchit/ansible-runner:kafka", "labels":map[string]interface {}{"name":"kafka-broker-network-latency"}, "permissions":map[string]interface {}{"apiGroups":[]interface {}{"", "extensions", "apps", "batch", "litmuschaos.io"}, "resources":[]interface {}{"daemonsets", "statefulsets", "deployments", "replicasets", "jobs", "pods", "pods/exec", "chaosengines", "chaosexperiments", "chaosresults"}, "verbs":[]interface {}{"*"}}}}}: validation failure list:
spec.definition.command in body must be of type array: "string"
ksatchit@playground:~/demo$ 
```

**Comment by ksatchit** [comment]
Example validation for chaosengine with pattern matching for spec items: 

```
The ChaosEngine "kafka-chaos" is invalid: []: Invalid value: map[string]interface {}{"apiVersion":"litmuschaos.io/v1alpha1", "kind":"ChaosEngine", "metadata":map[string]interface {}{"annotations":map[string]interface {}{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"litmuschaos.io/v1alpha1\",\"kind\":\"ChaosEngine\",\"metadata\":{\"annotations\":{},\"name\":\"kafka-chaos\",\"namespace\":\"default\"},\"spec\":{\"appinfo\":{\"appkind\":\"deployments\",\"applabel\":\"app=kafka\",\"appns\":\"default\"},\"chaosServiceAccount\":\"kafka-sa\",\"experiments\":[{\"name\":\"kafka-broker-pod-failure\",\"spec\":{\"components\":[{\"name\":\"KAFKA_REPLICATION_FACTOR\",\"value\":\"3\"},{\"name\":\"KAFKA_LABEL\",\"value\":\"app=kafka\"},{\"name\":\"KAFKA_NAMESPACE\",\"value\":\"default\"},{\"name\":\"KAFKA_SERVICE\",\"value\":\"kafka-svc\"},{\"name\":\"KAFKA_PORT\",\"value\":\"9092\"},{\"name\":\"KAFKA_CONSUMER_TIMEOUT\",\"value\":\"30000\"},{\"name\":\"KAFKA_INSTANCE_NAME\",\"value\":\"kafka\"},{\"name\":\"ZOOKEEPER_NAMESPACE\",\"value\":\"default\"},{\"name\":\"ZOOKEEPER_LABEL\",\"value\":\"app=zookeeper\"},{\"name\":\"ZOOKEEPER_SERVICE\",\"value\":\"zookeeper-instance-hs\"},{\"name\":\"ZOOKEEPER_PORT\",\"value\":\"2181\"},{\"name\":\"TOTAL_CHAOS_DURATION\",\"value\":\"60\"},{\"name\":\"CHAOS_INTERVAL\",\"value\":\"20\"},{\"name\":\"FORCE\",\"value\":\"false\"}]}}],\"jobCleanUpPolicy\":\"del\",\"monitoring\":false}}\n"}, "creationTimestamp":"2019-12-20T05:01:11Z", "generation":3, "name":"kafka-chaos", "namespace":"default", "resourceVersion":"16343720", "uid":"b5f0118c-9b77-4837-885f-52b1bcceda43"}, "spec":map[string]interface {}{"appinfo":map[string]interface {}{"appkind":"deployments", "applabel":"app=kafka", "appns":"default"}, "chaosServiceAccount":"kafka-sa", "experiments":[]interface {}{map[string]interface {}{"name":"kafka-broker-pod-failure", "spec":map[string]interface {}{"components":[]interface {}{map[string]interface {}{"name":"KAFKA_REPLICATION_FACTOR", "value":"3"}, map[string]interface {}{"name":"KAFKA_LABEL", "value":"app=kafka"}, map[string]interface {}{"name":"KAFKA_NAMESPACE", "value":"default"}, map[string]interface {}{"name":"KAFKA_SERVICE", "value":"kafka-svc"}, map[string]interface {}{"name":"KAFKA_PORT", "value":"9092"}, map[string]interface {}{"name":"KAFKA_CONSUMER_TIMEOUT", "value":"30000"}, map[string]interface {}{"name":"KAFKA_INSTANCE_NAME", "value":"kafka"}, map[string]interface {}{"name":"ZOOKEEPER_NAMESPACE", "value":"default"}, map[string]interface {}{"name":"ZOOKEEPER_LABEL", "value":"app=zookeeper"}, map[string]interface {}{"name":"ZOOKEEPER_SERVICE", "value":"zookeeper-instance-hs"}, map[string]interface {}{"name":"ZOOKEEPER_PORT", "value":"2181"}, map[string]interface {}{"name":"TOTAL_CHAOS_DURATION", "value":"60"}, map[string]interface {}{"name":"CHAOS_INTERVAL", "value":"20"}, map[string]interface {}{"name":"FORCE", "value":"false"}}}}}, "jobCleanUpPolicy":"del", "monitoring":false}}: validation failure list:
spec.appinfo.appkind in body should match '^(deployment|statefulset|daemonset)$'
spec.jobCleanUpPolicy in body should match '^(delete|retain)$'
```

**Comment by ksatchit** [comment]
Example validation for chaosexperiment with malformed configmap/secret spec: 

```
ksatchit@playground:~/demo$ kubectl apply -f kafka-broker-network-latency.yaml 
The ChaosExperiment "kafka-broker-network-latency" is invalid: []: Invalid value: map[string]interface {}{"apiVersion":"litmuschaos.io/v1alpha1", "description":map[string]interface {}{"message":"Deleting a kafka broker pod\n"}, "kind":"ChaosExperiment", "metadata":map[string]interface {}{"annotations":map[string]interface {}{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"litmuschaos.io/v1alpha1\",\"description\":{\"message\":\"Deleting a kafka broker pod\\n\"},\"kind\":\"ChaosExperiment\",\"metadata\":{\"annotations\":{},\"generation\":1,\"name\":\"kafka-broker-network-latency\",\"namespace\":\"default\"},\"spec\":{\"definition\":{\"args\":[\"-c\",\"ansible-playbook ./experiments/kafka/kafka-broker-network-latency/kafka-broker-network-latency-ansible-logic.yml -i /etc/ansible/hosts -vv; exit 0\"],\"command\":[\"/bin/bash\"],\"configmaps\":[{\"mountPath\":null,\"name\":null}],\"env\":[{\"name\":\"ANSIBLE_STDOUT_CALLBACK\",\"value\":\"default\"},{\"name\":\"KAFKA_KIND\",\"value\":\"statefulset\"},{\"name\":\"KAFKA_LIVENESS_STREAM\",\"value\":\"enabled\"},{\"name\":\"KAFKA_LIVENESS_IMAGE\",\"value\":\"litmuschaos/kafka-client:ci\"},{\"name\":\"KAFKA_CONSUMER_TIMEOUT\",\"value\":\"30000\"},{\"name\":\"TOTAL_CHAOS_DURATION\",\"value\":\"60000\"},{\"name\":\"TARGET_CONTAINER\",\"value\":\"k8skafka\"},{\"name\":\"NETWORK_LATENCY\",\"value\":\"60000\"},{\"name\":\"NETWORK_INTERFACE\",\"value\":\"eth0\"},{\"name\":\"KAFKA_INSTANCE_NAME\",\"value\":\"\"},{\"name\":\"KAFKA_NAMESPACE\",\"value\":\"\"},{\"name\":\"KAFKA_LABEL\",\"value\":\"\"},{\"name\":\"KAFKA_BROKER\",\"value\":\"\"},{\"name\":\"KAFKA_REPLICATION_FACTOR\",\"value\":\"\"},{\"name\":\"KAFKA_SERVICE\",\"value\":\"\"},{\"name\":\"KAFKA_PORT\",\"value\":\"\"},{\"name\":\"ZOOKEEPER_NAMESPACE\",\"value\":\"\"},{\"name\":\"ZOOKEEPER_LABEL\",\"value\":\"\"},{\"name\":\"ZOOKEEPER_SERVICE\",\"value\":\"\"},{\"name\":\"ZOOKEEPER_PORT\",\"value\":\"\"},{\"name\":\"LIB\",\"value\":\"pumba\"},{\"name\":\"LIB_IMAGE\",\"value\":\"gaiaadm/pumba:0.6.5\"}],\"image\":\"ksatchit/ansible-runner:kafka\",\"labels\":{\"name\":\"kafka-broker-network-latency\"},\"permissions\":{\"apiGroups\":[\"\",\"extensions\",\"apps\",\"batch\",\"litmuschaos.io\"],\"resources\":[\"daemonsets\",\"statefulsets\",\"deployments\",\"replicasets\",\"jobs\",\"pods\",\"pods/exec\",\"chaosengines\",\"chaosexperiments\",\"chaosresults\"],\"verbs\":[\"*\"]},\"secrets\":null}}}\n"}, "creationTimestamp":"2019-12-20T05:55:06Z", "generation":1, "name":"kafka-broker-network-latency", "namespace":"default", "uid":"d65c6f8d-b664-44f4-b6d1-be101673befa"}, "spec":map[string]interface {}{"definition":map[string]interface {}{"args":[]interface {}{"-c", "ansible-playbook ./experiments/kafka/kafka-broker-network-latency/kafka-broker-network-latency-ansible-logic.yml -i /etc/ansible/hosts -vv; exit 0"}, "command":[]interface {}{"/bin/bash"}, "configmaps":[]interface {}{map[string]interface {}{"mountPath":interface {}(nil), "name":interface {}(nil)}}, "env":[]interface {}{map[string]interface {}{"name":"ANSIBLE_STDOUT_CALLBACK", "value":"default"}, map[string]interface {}{"name":"KAFKA_KIND", "value":"statefulset"}, map[string]interface {}{"name":"KAFKA_LIVENESS_STREAM", "value":"enabled"}, map[string]interface {}{"name":"KAFKA_LIVENESS_IMAGE", "value":"litmuschaos/kafka-client:ci"}, map[string]interface {}{"name":"KAFKA_CONSUMER_TIMEOUT", "value":"30000"}, map[string]interface {}{"name":"TOTAL_CHAOS_DURATION", "value":"60000"}, map[string]interface {}{"name":"TARGET_CONTAINER", "value":"k8skafka"}, map[string]interface {}{"name":"NETWORK_LATENCY", "value":"60000"}, map[string]interface {}{"name":"NETWORK_INTERFACE", "value":"eth0"}, map[string]interface {}{"name":"KAFKA_INSTANCE_NAME", "value":""}, map[string]interface {}{"name":"KAFKA_NAMESPACE", "value":""}, map[string]interface {}{"name":"KAFKA_LABEL", "value":""}, map[string]interface {}{"name":"KAFKA_BROKER", "value":""}, map[string]interface {}{"name":"KAFKA_REPLICATION_FACTOR", "value":""}, map[string]interface {}{"name":"KAFKA_SERVICE", "value":""}, map[string]interface {}{"name":"KAFKA_PORT", "value":""}, map[string]interface {}{"name":"ZOOKEEPER_NAMESPACE", "value":""}, map[string]interface {}{"name":"ZOOKEEPER_LABEL", "value":""}, map[string]interface {}{"name":"ZOOKEEPER_SERVICE", "value":""}, map[string]interface {}{"name":"ZOOKEEPER_PORT", "value":""}, map[string]interface {}{"name":"LIB", "value":"pumba"}, map[string]interface {}{"name":"LIB_IMAGE", "value":"gaiaadm/pumba:0.6.5"}}, "image":"ksatchit/ansible-runner:kafka", "labels":map[string]interface {}{"name":"kafka-broker-network-latency"}, "permissions":map[string]interface {}{"apiGroups":[]interface {}{"", "extensions", "apps", "batch", "litmuschaos.io"}, "resources":[]interface {}{"daemonsets", "statefulsets", "deployments", "replicasets", "jobs", "pods", "pods/exec", "chaosengines", "chaosexperiments", "chaosresults"}, "verbs":[]interface {}{"*"}}, "secrets":interface {}(nil)}}}: validation failure list:
spec.definition.secrets in body must be of type array: "null"
spec.definition.configmaps.mountPath in body must be of type string: "null"
spec.definition.configmaps.name in body must be of type string: "null"
```

File: deploy/crds/chaosengine_crd.yaml [hunk 1]
```diff
@@ -28,6 +29,60 @@ spec:
...
+              properties: 
+                monitor: 
+                  type: object
+                  properties: 
+                    image: 
+                      type: string
+                runner: 
+                  type: object
+                  properties:
+                    image:
```
Comment (Sanjay1611) [inline deploy/crds/chaosengine_crd.yaml]: Should we also have a `type` field here to support `goExecutor`???

File: deploy/crds/chaosengine_crd.yaml [hunk 2]
```diff
@@ -28,6 +29,60 @@ spec:
...
+              type: array
+              items:
+                type: object
+                properties:
+                  name: 
+                    type: string
+                  spec:
+                    type: object
+                    properties:
+                      components:
```
[the remark this answers is not in the document]
Comment (ksatchit) [reply deploy/crds/chaosengine_crd.yaml]: not used currently, have omitted it from the validation as we don't have the supporting logic for this, neither do we have in examples.

File: deploy/crds/chaosengine_crd.yaml (hunk 1 above)
Comment (ksatchit) [reply deploy/crds/chaosengine_crd.yaml]: Yes. Thanks for pointing out. Adding this

Comment (ksatchit) [reply deploy/crds/chaosengine_crd.yaml]: Have added the validation for runner type!

Comment (Sanjay1611) [reply deploy/crds/chaosengine_crd.yaml]: Thanks @ksatchit

**Pull request merged**